src/main.rs

```diff
@@ -1,3 +1,131 @@
-fn main() {
-    println!("Hello, world!");
+use crate::config::read_config;
+use anyhow::Context;
+use hickory_proto::xfer::DnsResponse;
+
+use std::str::FromStr;
+use time::Duration;
+
+use hickory_client::{
+    client::Client,
+    op::ResponseCode,
+    rr::{rdata::TXT, Name, RData, Record, RecordType},
+};
+
+use log::{debug, info, trace};
+extern crate pretty_env_logger;
+#[macro_use]
+extern crate log;
+
+mod cli;
+mod config;
+mod happy;
+
+fn main() -> anyhow::Result<()> {
+    pretty_env_logger::init_timed();
+
+    let cli = cli::parse();
+
+    let config = read_config(&cli.config)?;
+    debug!("initializing with zones {:?}", config.zones.keys());
+
+    let challenge_name: Name = find_record(cli.identifier)?;
+    info!("challenge name is {}", challenge_name);
+
+    let zone: Name = find_zone(challenge_name.clone())?;
+    info!("challenge zone is {}", zone);
+
+    let primary_ns_client = &config
+        .zones
+        .get(&zone)
+        .context("Couldn't find challenge zone in config")?
+        .create_client();
+
+    let challenge_record = Record::from_rdata(
+        challenge_name.clone(),
+        Duration::minutes(1).whole_seconds() as u32,
+        RData::TXT(TXT::new(Vec::from([cli.proof]))),
+    );
+
+    match &cli.command.unwrap() {
+        cli::Commands::Set => {
+            let result = primary_ns_client
+                .create(challenge_record.clone(), zone)
+                .context("Couldn't set record")?;
+            match result.response_code() {
+                ResponseCode::NoError => {
+                    info!("record was successfully set");
+                    Ok(())
+                }
+                ResponseCode::YXRRSet => {
+                    warn!("{:?} already exists", challenge_record.clone());
+                    Ok(())
+                }
+                _ => todo!(),
+            }
+        }
+        cli::Commands::Cleanup => {
+            let result = primary_ns_client
+                .delete_rrset(challenge_record.clone(), zone)
+                .context("Couldn't remove record")?;
+            match result.response_code() {
+                ResponseCode::NoError => {
+                    info!("record was removed");
+                    Ok(())
+                }
+                _ => {
+                    debug!("{:?}", result);
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+fn find_record(identifier: Name) -> anyhow::Result<Name> {
+    let prefix: Name = Name::from_str("_acme-challenge").unwrap();
+    let q = happy::search(
+        prefix.append_name(&identifier)?, // this should be called prepend because that's what it actualyl does
```

Comment:
Well, it takes the Name created above and appends an identifier to it, so append is fine? prepend would be `identifier.prepend_name(prefix)`, no?